xdsl/utils/disjoint_set.py

```diff
@@ -0,0 +1,170 @@
+"""
+Generic implementation of a disjoint set data structure.
+
+https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+"""
+
+from collections.abc import Hashable, Sequence
+from typing import Generic, TypeVar
+
+
+class IntDisjointSet:
+    """
+    Represents a collection of disjoint sets of integers.
+    The integers stored are always in the range [0,n), where n is the number of elements
+    in this structure.
+
+    This implementation uses path compression and union by size for efficiency.
+    The amortized time complexity for operations is nearly constant.
+    """
+
+    _parent: list[int]
+    """
+    Index of the parent node. If the node is its own parent then it is a root node.
+    """
+    _count: list[int]
+    """
+    If the node is a root node, the corresponding value is the count of elements in the
+    set. For non-root nodes, these counts may be stale and should not be used.
+    """
+
+    def __init__(self, *, size: int) -> None:
+        """
+        Initialize disjoint sets with elements [0,size).
+        Each element starts in its own singleton set.
+        """
+        self._parent = list(range(size))
+        self._count = [1] * size
+
+    def value_count(self) -> int:
+        """Number of nodes in this structure."""
+        return len(self._parent)
+
+    def set_count(self) -> int:
+        """
+        Number of sets in this structure.
+        Note: This is O(n) as it needs to scan all parents.
+        """
+        return len(set(self._parent))
```

Comment:
Parent pointers are not eagerly updated so I don't think this is correct? If this operation is needed then it's likely better to keep track of the number of sets in a variable

Reply:
Good point I'll remove it